### PR TITLE
Avoid optional `Tracker` attributes and enable mypy

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

Refactor the progress structure to avoid the `Optional[int]` attributes. They made typing these files pretty tricky.

We had it previously to avoid users accidentally setting unused attributes. We now use separate classes so the attributes don't even exist.

Part of #5023

### Does your PR introduce any breaking changes? If yes, please list them.

None user-facing.

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

### CODE REVIEW

**Review by tchaton — APPROVED** [review]

**Review by Borda — APPROVED** [review]

File: tests/loops/test_loop_state_dict.py [hunk 1]
```diff
@@ -53,28 +53,25 @@ def test_loops_state_dict_structure():
...
+                        "total": {"ready": 0, "started": 0, "completed": 0},
+                        "current": {"ready": 0, "started": 0, "completed": 0},
                     },
                 },
                 "optimizer_idx": 0,
             },
             "epoch_loop.val_loop.state_dict": {},
             "epoch_loop.val_loop.dataloader_progress": {
-                "total": {"ready": 0, "started": None, "processed": None, "completed": 0},
-                "current": {"ready": 0, "started": None, "processed": None, "completed": 0},
```
Comment (awaelchli) [inline tests/loops/test_loop_state_dict.py:77]: nice that these disappear. 
will loading the "old checkpoints" be agnostic to this change? but right now we are only storing it for the FT case right? so it's fine I guess :)

**Review by awaelchli — APPROVED** [review]

File: tests/loops/test_loop_state_dict.py (hunk 1 above)
Comment (carmocca) [reply tests/loops/test_loop_state_dict.py:77]: Correct!
